server/utils/utils.go

```diff
@@ -0,0 +1,330 @@
+		if err != nil {
+			return "", err
+		}
+	}
+	if u.Host == "" {
+		return "", errors.Errorf("Invalid URL, no hostname: %q", confluenceURL)
+	}
+	if u.Scheme == "" {
+		u.Scheme = "https"
+	}
+
+	confluenceURL = strings.TrimSuffix(u.String(), "/")
+	return confluenceURL, nil
+}
+
+// CheckConfluenceURL checks if `/status` endpoint of the Confluence URL is accessible
+// and responding with the correct state which is "RUNNING"
+func CheckConfluenceURL(mattermostSiteURL, confluenceURL string, requireHTTPS bool) (_ string, err error) {
+	confluenceURL, err = NormalizeConfluenceURL(confluenceURL)
+	if err != nil {
+		return "", err
+	}
+	if confluenceURL == strings.TrimSuffix(mattermostSiteURL, "/") {
+		return "", errors.Errorf("%s is the Mattermost site URL. Please use your Confluence URL", confluenceURL)
+	}
+
+	defer func() {
+		if err != nil {
+			err = errors.Wrap(err, "we couldn't validate the connection to your Confluence server. "+
+				"This could be because of existing firewall or proxy rules, or because the URL was entered incorrectly")
+		}
+	}()
+
+	resp, err := http.Get(confluenceURL + "/status")
+	if err != nil {
+		return "", err
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		return "", errors.Errorf("Confluence server returned http status code %q when checking for availability: %q", resp.Status, confluenceURL)
+	}
+
+	resBody, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return "", err
+	}
+	var status ConfluenceStatus
+	err = json.Unmarshal(resBody, &status)
+	if err != nil {
+		return "", err
+	}
+	if status.State != "RUNNING" {
+		return "", errors.Errorf("Confluence server is not in correct state, it should be up and running: %q", confluenceURL)
+	}
+	return confluenceURL, nil
+}
+
+func IsConfluenceCloudURL(confluenceURL string) bool {
+	u, err := url.Parse(confluenceURL)
+	if err != nil {
+		return false
+	}
+	return strings.HasSuffix(u.Hostname(), ".atlassian.net")
+}
+
+func CallJSONWithURL(instanceURL, path, method string, in, out interface{}, httpClient *http.Client) (responseData []byte, err error) {
+	urlPath, err := GetEndpointURL(instanceURL, path)
+	if err != nil {
+		return nil, err
+	}
+
+	return CallJSON(instanceURL, method, urlPath, in, out, httpClient)
+}
+
+func GetEndpointURL(instanceURL, path string) (string, error) {
+	endpointURL, err := url.Parse(strings.TrimSpace(fmt.Sprintf("%s%s", instanceURL, path)))
+	if err != nil {
+		return "", err
+	}
+
+	return endpointURL.String(), nil
+}
+
+func CallJSON(url, method, path string, in, out interface{}, httpClient *http.Client) (responseData []byte, err error) {
+	contentType := "application/json"
+	buf := &bytes.Buffer{}
+	err = json.NewEncoder(buf).Encode(in)
+	if err != nil {
+		return nil, err
+	}
+	return call(url, method, path, contentType, buf, out, httpClient)
+}
+
+func call(basePath, method, path, contentType string, inBody io.Reader, out interface{}, httpClient *http.Client) (responseData []byte, err error) {
```

Comment:
Can we separate this function so it only does one thing? Could we handle/manage the response in another function and to encapsulate the response and make it easier to read and test in the future? from line 246 until around 275/280 I think?